examples/Reranking_search_results_with_cross-encoders.ipynb

```diff
@@ -0,0 +1,805 @@
+      "19: Efficient and Flexible Topic Modeling using Pretrained Embeddings and Bag of Sentences\n",
+      "20: Relational Sentence Embedding for Flexible Semantic Matching\n"
+     ]
+    }
+   ],
+   "source": [
+    "for i, result in enumerate(result_list):\n",
+    "    print(f\"{i + 1}: {result['title']}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d5727678",
+   "metadata": {},
+   "source": [
+    "## Cross-encoder\n",
+    "\n",
+    "We'll create a cross-encoder using the ```Completions``` endpoint - the key factors to consider here are:\n",
+    "- Make your examples domain-specific - the strength of cross-encoders comes when you fine-tune them to your domain.\n",
+    "- There is a trade-off between how many potential examples to re-rank vs. processing speed. Consider batching and parallel processing cross-encoder requests to aid user experience.\n",
+    "\n",
+    "The steps here are:\n",
+    "- Build a prompt to assess relevance and provide few-shot examples to tune it to your domain\n",
+    "- Return the classification of yes/no as well as the ```logprobs```\n",
+    "- Rerank the results by the ```logprobs``` keyed on Yes\n",
+    "- Add a ```logit bias``` for the tokens for Yes and No to decrease the likelihood of any other tokens occurring"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "id": "ca634bf9",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "([5297], [2949])"
+      ]
+     },
+     "execution_count": 6,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "tokens = ['Yes','No']\n",
+    "tokenizer = tiktoken.get_encoding(\"gpt2\")\n",
+    "ids = [tokenizer.encode(token) for token in tokens]\n",
+    "ids[0],ids[1]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "id": "4fdf8c11",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "prompt = '''\n",
+    "You are an Assistant responsible for helping detect whether the retrieved document is relevant to the query. For a given input, you need to output a single token: \"Yes\" or \"No\" indicating the retrieved document is relevant to the query.\n",
+    "\n",
+    "Examples:\n",
+    "\"\"\"\n",
+    "Query: How to plant a tree?\n",
+    "Retrieved document: Cars were invented in 1886, when German inventor Carl Benz patented his Benz Patent-Motorwagen.[3][4][5] Cars became widely available during the 20th century. One of the first cars affordable by the masses was the 1908 Model T, an American car manufactured by the Ford Motor Company. Cars were rapidly adopted in the US, where they replaced horse-drawn carriages.[6] In Europe and other parts of the world, demand for automobiles did not increase until after World War II.[7] The car is considered an essential part of the developed economy.\n",
```

Comment:
optional: you can potentially wrap these strings in delimiters (e.g., `"""`). Otherwise it might get less reliable on documents or queries that have newlines and colons within them.